### PR TITLE
Reference resolution on ingest for faster chained searches

### PR DESCRIPTION
**Pull request by johnstairs**
## Description

We currently defer the resolution of resource references to search time. This PR changes that to perform and maintain it on ingestion. We add a `ResolvedResourceSurrogateId` to the `ReferenceSearchParam` table and indexes so that the field can be read and updated efficiently. 

The result is a improvement in performance for chained searches. This PR does not address `_include` performance.

The plan was to use an indexed view but performance hit was a lot higher than maintaining this column ourselves.

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/10.diff.sql [hunk 1]
```diff
@@ -0,0 +1,585 @@
+SET NOCOUNT ON
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/10.diff.sql:1]: >SET [](http://example.com/codeflow?start=0&length=3)

I can probably do comparison between 9 and 10 version, and then between 10 and 11, and try to deduct why 2 versions needed, but I better ask directly.

Most of the time it's one time migration per PR, why it's double migration this time?

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/10.diff.sql:1]: First one ensures that new data will have the new column set (the stored procedure writes is). The second one updates all the existing data. When the second is done, you can start reading from the column.

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/10.diff.sql:1]: Actually, I guess you could combine them.

**Comment by CaitlinV39** [comment]
We are going to close this for now based on priorities but will leave the code for reference for when we have time to revisit.

**Pull request closed without merging**